### PR TITLE
Support defensive hardpoint setting that doesn't affect missile boat AI mode

### PR DESCRIPTION
**Pull request by warp-core**
**Feature:** This PR implements the feature request detailed and discussed in issue #6712

## Feature Details
Hardpoints can be set to 'defensive'. When determining the shortest range, AI::Attack will ignore weapons in defensive hardpoints until non-defensive weapons can't be fired (if there is no ammo for them).
This is not done yet, I've only taken the first steps.
Need to allow the game to read and write 'defensive' marks to hardpoints with weapons in the save and data files and need a way for the ShipInfoPanel to change the defensiveness of a hardpoint.
Also going to allow 'opportunistic' to be defined on a per hardpoint basis, so defensive hardpoints on a missile boat can operate opportunistically and attack ships that manage to get close without the ship having to cease its assault on its main target at range with its standoff weapons, until it decides the nearer ship is a greater threat and deserves the attention of its main weapons, perhaps.

Please tell me if what I'm doing is terrible.

## UI Screenshots
This PR does not make any UI changes.

~~The change to ShipInfoPanel::DrawWeapons() is supposed to be temporary, just for testing, I'm not sure what the method for conveying the new hardpoint properties should be.
Before:
image
After:
image~~

## Usage Examples
```
turret 10 50 "Korath Banisher"
		defensive
		opportunistic
```
Use one, both or neither. Works in variant definitions where the hardpoint coordinates aren't present.
```
turret "Korath Banisher"
		defensive
		opportunistic
```
Does not (or at least, should not) work when no outfit is specified.

## Testing Done
[defensiveandopportunisticships.txt](https://github.com/endless-sky/endless-sky/files/8536099/defensiveandopportunisticships.txt) - should be placed in the data folder of a plugin or the game files themselves.
[DefensiveOpportunisttic Test.txt](https://github.com/endless-sky/endless-sky/files/8536101/DefensiveOpportunisttic.Test.txt) - save file.
I've increased the lifetime of the Detainer from 140 to 150 so its range exceeds 2000 and it can allow missile boat AI without being defensive.
The first of the 78s has no defensive weapons. Will charge in even if you remove the Detainer because of the Banisher.
The second has a defensive Banisher and so does not charge in as aggressively, though still gets quite close, I assume because the Banisher is not homing and so the AI considers the range to be relatively short. Removing the Detainer makes it behave like the third ship despite having a Banisher (thanks to that Banisher being defensive).
The third only has long range, homing missile launchers and anti-missile turrets, and so will operate under missile boat AI with the current code already.

## Performance Impact
{{describe any performance impact (positive or negative). "N/A" if no performance-critical code is changed. }}
Untested, I don't expect it to be huge but the changes I've made are definitely not improving performance.

### CODE REVIEW

**Comment by warp-core** [comment]
So, it should now be able to recognise 'defensive' and 'opportunistic' child tokens for hardpoints in ship definitions.
These will be discarded if the hardpoint doesn't have a weapon in it.
Ship::Save() will also write the appropriate token(s) if the (occupied) hardpoint is marked as defensive and/or opportunistic.
It should not currently be possible for an unoccupied hardpoint to be marked as defensive or opportunistic and I think it should remain this way.
Actually, I probably need to have Hardpoint::Uninstall() set those to false.

**Comment by warp-core** [comment]
Ok, that's done.

**Comment by Pointedstick** [comment]
Interesting approach. I wonder if a better and more automatic way could be to compare the DPS from a ship's turrets to the DPS from its weapons requiring ammunition. If the latter is greater than the former, always use the missile boat AI. And if the ship has any ammo-less guns, it never uses the missile boat AI.

**Comment by warp-core** [comment]
That would certainly be more automatic, but allowing it to be set explicitly allows much more flexibility.
Ammo-less guns disabling missile boat mode seems undesirable. Detainers are ammo-less guns and yet seem pretty perfect for missile boat behaviour (not sure if the 1960 range is long enough?).

**Comment by Pointedstick** [comment]
Ammo-less guns with 0 or low DPS could also be ignored, maybe. Like you'd be a missile boat if your turret DPS is less than your missile DPS, or your gun DPS is less than 1/4 your missile DPS. Something like that. But your point is fair. I'm not opposed to making it configurable, but I think a good automatic default is nice too. Usually nicer, if you can get the conditions right. :)

**Comment by Zitchas** [comment]
I agree, an automatic option would be nice; but I definitely want to be able to manually set things.

One thing that occurs to me: It would be good if the manual setting could be set up in ship definitions. Being able to setup stock and variant ship definitions something like this:

```
		turret -7 -4 "Electron Turret"
		turret 7 -4 "Electron Turret"
		turret -7 -4 "Heavy Anti-Missile Turret"
		turret 7 -4 "Heavy Anti-Missile Turret"
		turret -23 14 "Electron Turret"
			defensive
		turret 23 14 "Electron Turret"
			defensive
```
Being able to setup the ships in NPC fleets with proper "this is considered a defensive weapon for this ship" specifications would be awesome. (for instance, if we had a carrier or artillery ship with some proper Auroran Carrier style ultra-long range artillery, even heavy-but-short-range turrets may be best treated as a defensive turret.)

**Comment by warp-core** [comment]
That's exactly what I'm going for. If I've modified Ship::Load() and Ship::Save() properly, it should be able to read and save tokens of that form.

**Comment by warp-core** [comment]
I don't know what's happening at AI.cpp lines 1261 and 2101, are they whitespace issues?

**Comment by Saturnalia9** [comment]
So not entirely sure what's going on (and no idea how to show what's happening), but my turrets all seem to be turning away from the targets at odd angles, rather than aiming toward when I loaded up the save and launched. I didn't change any settings.

**Comment by warp-core** [comment]
That might mean I broke something.
In your settings, do you have turret tracking set to focused it opportunistic?

**Comment by warp-core** [comment]
Updated the OP.
Next steps are to work on allowing the player to change these settings in the ship info panel and displaying this information there.

**Comment by warp-core** [comment]
Is 'ready for review' correct? It's definitely not ready for merging but I want feedback.

**Comment by petervdmeer** [comment]
The option to set "opportunistic" mode per turret (or per hardpoint) was also requested in https://github.com/endless-sky/endless-sky/issues/4714

**Comment by petervdmeer** [comment]
I'm also not too happy about the word `"defensive"`, since we are actually controlling which weapon is preferred for active attacking. If we want to manually specify per hardpoint which weapon is used when during attack, then I could imagine that we use a keyword like `"attack preference"`, where 0 means; never use for active attack, and higher numbers indicate a preference for an AI to use said weapon above lower numbered weapon. Using such a mechanism allows to have more levels of preferred attack weapons than just 2.

And I wonder if this is best specified as a per-turret property or if this should be a "personality". If we only want to differentiate between "attacking with long range weapons only" and "attacking with all weapons we have", then a personality seems more fitting than settings per hardpoint.

**Comment by Zitchas** [comment]
From a content-creation (as well as player fleet management) perspective, I think this fits best on the "per turret" format.

1. Per turret allows us to specify exactly which turrets the AI is supposed to use to determine engagement distance.
2. Per turret allows players to specify which of the turrets on ships in their fleet the AI uses. 
3. Shifting this to a personality means that we need to establish broad rules or patterns for identifying which weapons it should be using to determine engagement range, and then ensuring that configurations always match the pattern. 
4. Personality based determination removes this from player utilization unless we add a player-visible "AI selector" that they can use to assign AI behaviors to ships in their fleet. I think this would be out-of-scope for this PR, but could also be very well received. I virtually always would like to give my escorts the "coward" personality, for instance.

The typical scenario for making use of this would be ships that have a long-range heavy hitting weapon, and then a short-range lighter/faster weapon for defensive purposes. 

That being said, we could just as easily have ships that have three or four different ranges of weapons. Perhaps the ship has missiles or a lightweight but very long-range weapon as a hyperjump disruption weapon as well as mid-range heavy turrets and short range anti-fighter turret. Depending on the loadout, we might want to have the ship prioritize engagement based on the mid-range heavy turret, and ignore both the longer range missiles & light-weight tackling turret as well as the close range AAA.

At the end of the day, I disliking handing over control of strategic determination of weapons usage to AI, and wherever possible prefer to be able to tell the ship specifically *this* is your primary weapon, only adjust combat distance if to the others if this primary weapon is out of ammo or the enemy closes to a short enough distance that the others are in range.

**Comment by Saturnalia9** [comment]
> That might mean I broke something. In your settings, do you have turret tracking set to focused it opportunistic?

At the time, I had it set to focused. It was still attempting to fire to the opposite. I'll give it another try now.

**Comment by warp-core** [comment]
I'm considering if it would be better to make UI changes to allow players to change the opportunistic/defensiveness of weapons on their own ships in a separate PR, this PR just adds the AI changes to support the new features and allows ship definitions to use the new attributes.

**Comment by warp-core** [comment]
Looking at #4714 as @petervdmeer suggested, this PR should allow 'opportunistic' to be a per hardpoint property.
Currently, the default is non-opportunistic (uninstalling a turret will falsify the opportunistic attribute, if `opportunistic` is not explicitly stated in the ship definition, it will be false, an empty hardpoint cannot be set to opportunistic). Is the general desire for the default to be opportunistic?
Additionally, the turret tracking preference currently overrides the per hardpoint settings if it is set to opportunistic (all turrets will be opportunistic regardless of the player's per hardpoint settings) but not if it is set to focused (the player's settings will be used). I suppose I should probably change the 'focused' label on that preference to something else.

https://github.com/endless-sky/endless-sky/issues/4714#issuecomment-565163575
I think it shouldn't be too much more work to add this option for guns, and support it in #6822.

There's some more UI stuff mentioned there which I'll bring up in #6822.

**Comment by Zitchas** [comment]
I'd say that the fundemental baseline default behavior should probably be focused. As in, they focus on whatever the player has targeted. 

That being said, I would really appreciate being able to set a turret-specific default behavior. So while most turrets default to focused, the ones that specifically call out their use as anti-fighter platforms (laser turret, blaster turret, and quad blaster turret for starters) should have an opportunistic targeting as their default.

As an aside that is out of scope for this PR; turret variation is in a rather bad place, I think. Looking through them trying to find sensible patterns and they just... don't. Things like modified blaster turret, which sounds like an improved heavy hitter version, is also faster at turning than the quad blaster turret, which sounds like the ideal "fill the sky with shots to take down fighters." Argh. Frustrating. But not your fault, or something for this PR.

**Comment by warp-core** [comment]
Current behaviour under this PR:
Turret tracking preference set to opportunistic or opportunistic personality present: all turrets behave opportunistically, should match the behaviour before this PR.
Turret tracking set to focused (opportunistic personality not present): individual turret behaviour determined by hardpoint setting. If it is opportunistic, it will behave opportunistically, if not, it will be focused.
Should the opportunistic preference be removed altogether?

File: source/Ship.cpp [hunk 1]
```diff
@@ -541,17 +547,45 @@ void Ship::FinishLoading(bool isNewInstance)
...
 				else
 				{
 					while(nextTurret != end && !nextTurret->IsTurret())
 						++nextTurret;
-					const Outfit *outfit = (nextTurret == end) ? nullptr : nextTurret->GetOutfit();
-					merged.AddTurret(bit->GetPoint() * 2., bit->IsUnder(), outfit);
+					const Outfit *outfit;
+					bool defensive;
+					bool opportunistic;
+					if (nextTurret == end)
```
Comment (samrocketman) [inline source/Ship.cpp:576]:
```suggestion
					if(nextTurret == end)
```

**Comment by petervdmeer** [comment]
> @petervdmeer have you had a chance to look much at the code?

Yes, I already briefly looked at the code before I gave my previous review comments; but if I have concerns with the overall direction, then I feel it only distracts if I give detailed review-comments on the code.

> Do you object to 'defensive' being used internally and in the data files? As you've seen, #6822 presents it as 'offensive' in the UI.

Yes, I object to that. The data-files are an "external interface" where content-creators specify what they want to see happening. I feel that such an external interface needs to use well-defined and non-ambiguous keywords.
If a content creator sets`"defensive"` to `"off"` for a ships-weapon in a data-file, and the weapon is still used for defending the ship, then I consider that ambiguous and confusing.

> Since 'offensive' is the default (meaning, the range of a weapon will be considered when determining missile boat behaviour unless explicitly marked as 'defensive'), removing not using 'defensive' in the ship definitions might be a little odd, what would the tage be? 'not-offensive'? Unless the default should be 'defensive' and weapons should only be prioritised if marked as 'offensive'?

`"not-offensive"` would be okay for me; the weapon would just not be considered for offensive operations. (But maybe you should also ask a few other content-creators, because I sometimes suggest keywords that are a little bit too complex.)

> I can change 'defensive' to 'offensive' in the code, if you think that'd be better?

For code-maintainability I feel it is very appropriate to use the same terminology in the data-files, in the code and in the UI (wherever possible).

**Comment by warp-core** [comment]
> > Do you object to 'defensive' being used internally and in the data files? As you've seen, #6822 presents it as 'offensive' in the UI.
> 
> Yes, I object to that. The data-files are an "external interface" where content-creators specify what they want to see happening. I feel that such an external interface needs to use well-defined and non-ambiguous keywords. If a content creator sets`"defensive"` to `"off"` for a ships-weapon in a data-file, and the weapon is still used for defending the ship, then I consider that ambiguous and confusing.
> 
> > Since 'offensive' is the default (meaning, the range of a weapon will be considered when determining missile boat behaviour unless explicitly marked as 'defensive'), removing not using 'defensive' in the ship definitions might be a little odd, what would the tage be? 'not-offensive'? Unless the default should be 'defensive' and weapons should only be prioritised if marked as 'offensive'?
> 
> `"not-offensive"` would be okay for me; the weapon would just not be considered for offensive operations. But maybe you should also ask a few other content-creators, because I sometimes suggest keywords that are a little bit too complex. Zitchas or Amazinite usually have pretty good ideas about fitting keywords.

@Zitchas @Amazinite thoughts? Preferences?

**Comment by Zitchas** [comment]
Sorry for missing this. 

I'm good with `"not-offensive"`, as that is very specific and to-the-point about what we are specifying here: Namely, that the weapon in this hardpoint is not considered when the AI is determining what range they should be aiming for.

It's not exactly the most streamlined or concise label, but I value precision and clarity a lot more than concision.

**Comment by Zitchas** [comment]
We're not getting this after all?

**Pull request closed without merging**